keras/src/layers/reshaping/reshape_test.py

```diff
@@ -100,6 +102,32 @@ def test_reshape_with_dynamic_batch_size_and_minus_one(self):
         reshaped = backend.compute_output_spec(layer.__call__, input)
         self.assertEqual(reshaped.shape, (None, 3, 8))
 
+    def test_reshape_layer_with_varying_input_size_and_minus_one(self):
+        input = KerasTensor((None, 6, 4))
+        layer = layers.Reshape((-1, 8))
+        layer.build(input.shape)
+        res = layer(ops.ones((1, 6, 4), dtype="float32"))
+        self.assertEqual(res.shape, (1, 3, 8))
+        res = layer(ops.ones((1, 10, 4), dtype="float32"))
+        self.assertEqual(res.shape, (1, 5, 8))
+
+    def test_custom_reshape_model_with_varying_input_size_and_minus_one(self):
```

Comment:
Let's do a compiled example instead, which exercise a different use case than the example right before. Also, the model can be just the `Reshape` layer:

```python
    def test_reshape_model_fit_with_varying_input_size_and_minus_one(self):
        def generator():
            yield (
                ops.ones((1, 12, 2), dtype="float32"),
                ops.zeros((1, 3, 8), dtype="float32"),
            )
            yield (
                ops.ones((1, 20, 2), dtype="float32"),
                ops.zeros((1, 5, 8), dtype="float32"),
            )

        layer = layers.Reshape((-1, 8))
        model = models.Sequential([layer])
        model.compile(loss="mean_squared_error")
        model.fit(generator())
```